build/src/lib.rs

```diff
@@ -1,26 +1,69 @@
 //! This crate provides a builder to make it easier to use the correct binaries for [nlprule](https://github.com/bminixhofer/nlprule).
 //! See `README.md` for details.
 
-use flate2::read::GzDecoder;
+use flate2::bufread::GzDecoder;
+use fs::File;
+use fs_err as fs;
 use nlprule::{compile, rules_filename, tokenizer_filename};
+use reqwest::blocking::Response;
+use std::fs::Permissions;
 use std::{
-    fs::{self, File},
-    io::{self, BufWriter, Cursor, Read},
+    io::{self, BufReader, BufWriter, Cursor, Read, Write, Seek, SeekFrom},
     path::{Path, PathBuf},
+    result,
 };
-use thiserror::Error;
 use zip::result::ZipError;
 
-#[derive(Debug, Error)]
+pub type OtherError = Box<dyn std::error::Error + Send + Sync + 'static>;
+
+#[derive(Debug, thiserror::Error)]
 pub enum Error {
-    #[error("request error: {0}")]
+    #[error(transparent)]
     RequestError(#[from] reqwest::Error),
-    #[error("i/o error: {0}")]
+    #[error("Binaries were not found on the remote")]
+    BinariesNotFound,
+    #[error("Failed to validate {1:?} binary for lang {0}")]
+    ValidationFailed(String, Binary, #[source] nlprule::Error),
+    #[error(transparent)]
     IOError(#[from] io::Error),
-    #[error("zip error: {0}")]
+    #[error(transparent)]
     ZipError(#[from] ZipError),
     #[error("error postprocessing binaries: {0}")]
-    PostprocessingError(Box<dyn std::error::Error>),
+    PostprocessingError(#[source] OtherError),
+    #[error("error postprocessing binaries: {0}")]
+    TransformError(#[source] OtherError),
+    #[error("Collation failed")]
+    CollationFailed(#[source] nlprule::compile::Error)
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+/// Definition of the data transformation for the network retrieved, binencoded rules and tokenizer datasets.
+pub trait TransformDataFn:
```

Comment:
Why does this need to take a `dyn Read` and `dyn Write`? Don't we always pass a `Cursor<Vec<u8>>` and `Cursor<&mut Vec<u8>>` to it? Changing this in the signature would simplify things quite a bit.

Reply:
Unfortunately that's not the case with the two incarnations when using it for the downloaded files or the spawned files, hence the complications. A rather big refactoring could resolve this, but I would prefer to do that separately.

Reply:
Okay sure! Let's keep this as it is. I might be missing something regarding how big the refactor would have to be. I might try this after the PR is merged. It is perfectly valid as it is too.